### PR TITLE
[mlir][linalg] Enable fuse consumer

### PR DESCRIPTION
**Pull request by cxy-1993**
This patch adds support for consumer fusion to the tiling interface, and implements fuse consumers on FuseIntoContainingOp.

- Add interface method 'getIterDomainTilePositionFromOperandPosition' to tiling interface which get iteration domain position from operand position.
- Add interface method 'getTiledImplementationFromOperandPosition' to tiling interface which generate tiled implementation according to operand position.
- Implemented the above two methods and supported consumer fusion for FuseIntoContainingOp.

### CODE REVIEW

**Comment by cxy-1993** [comment]
@MaheshRavishankar Could you please help me review this patch?

**Comment by MaheshRavishankar** [comment]
Thanks! Yes, I will definitely review the patch. I am in a middle of a crunch week. I'll probably get to it in about 10 days if that is ok

**Comment by cxy-1993** [comment]
> Thanks! Yes, I will definitely review the patch. I am in a middle of a crunch week. I'll probably get to it in about 10 days if that is ok

Don't worry, there's no rush. Take your time.

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (MaheshRavishankar) [inline mlir/include/mlir/Interfaces/TilingInterface.td:79]: I think this isnt as clear. Please see the description for `getResultTilePosition` maybe something closer to that (or improve both if you have a better description).

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (MaheshRavishankar) [inline mlir/include/mlir/Interfaces/TilingInterface.td:83]: Nit: Maybe rename to `getIterationDomainTileFromOperandTile`?

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (MaheshRavishankar) [inline mlir/include/mlir/Interfaces/TilingInterface.td:123]: Nit : 

```
Generates the IR that computes the tiled implementation of an operation based on
the operand tile

```

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (MaheshRavishankar) [inline mlir/include/mlir/Interfaces/TilingInterface.td:77]: Nit: Just a matter of reading. At least for me, fusing producer with consumer is more natural and is then easy to make the leap to fuse consumer with producer. So maybe position these methods textually after the corresponding methods that deal with the producer -> consumer fusion.

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (MaheshRavishankar) [inline mlir/include/mlir/Interfaces/TilingInterface.td:129]: Nit: 
```
This method generates a tiled implementation of the operation based on the tile of the operand required.
```

File: mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp
Comment (MaheshRavishankar) [inline mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp:135]: Nit: just make the operand `LinalgOp linalgOp` type instead of `Operation *op`. You can then do 

```
auto tilingInterfaceOp = cast<TilingInterface>(linalgOp.getOperation());
```

**Review by MaheshRavishankar — DISMISSED** [review]
Thanks for the changes. I looked at the changes to the `TilingInterface`... those look good. Just left a few nits.

I missed this, but could you instead look at the implementation in https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/SCF/Transforms/TileUsingInterface.cpp instead. That file does tile consumer and fuse producer using the `LoopLikeOpInterface` and works for both `scf.for` and `scf.forall`. The path that is here in `LinalgTransforms` is supposed to be deprecated (by me when I get some time). Last time I touched it, I made things fairly modular and readable (afaik) and just following that but instead fusing with consumers would be a great help. If added here, it will just be on my list of things to deprecate and move it to the above file anyway.

I havent looked at the `LinalgTransforms.*` changes (cause I am also not an expert in transform dialect). I think it would be better to make these methods available outside of transform dialect to be callable as standalone utility functions (like is done in the implementation I linked to above).

File: mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp
Comment (cxy-1993) [reply mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp:135]: Thanks for your comments! It's crunch week for me now, I'll address your review comments ASAP.

**Comment by MaheshRavishankar** [comment]
Hi, some of this was used in https://github.com/llvm/llvm-project/pull/88712 . THere are two things needed here..

1) We need the interface methods
2) We need the base implementation that can fuse consumers for both scf.for and scf.forall (and hopefully unified under a `LoopLikeOpInterface`)
3) We need the transform dialect ops.

Two co-ordinate between these two PRs, can we first land the interface methods. The implementation of (2) here for just scf.forall has some of the same issues that I highlighted in the other PR... You can wait for that to land (and help contribute/review) and then add (3). So can we split this PR first to just land the interface methods?

**Comment by cxy-1993** [comment]
> Hi, some of this was used in #88712 . THere are two things needed here..
> 
> 1. We need the interface methods
> 2. We need the base implementation that can fuse consumers for both scf.for and scf.forall (and hopefully unified under a `LoopLikeOpInterface`)
> 3. We need the transform dialect ops.
> 
> Two co-ordinate between these two PRs, can we first land the interface methods. The implementation of (2) here for just scf.forall has some of the same issues that I highlighted in the other PR... You can wait for that to land (and help contribute/review) and then add (3). So can we split this PR first to just land the interface methods?

Thank you for your review and reminders. I have already extracted the interface part separately in this MR submission.

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (cxy-1993) [reply mlir/include/mlir/Interfaces/TilingInterface.td]: DONE rename to getIterationDomainTileFromOperandTile

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (cxy-1993) [reply mlir/include/mlir/Interfaces/TilingInterface.td]: DONE

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (cxy-1993) [reply mlir/include/mlir/Interfaces/TilingInterface.td]: DONE, move methods after profucer fusion

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (cxy-1993) [reply mlir/include/mlir/Interfaces/TilingInterface.td]: DONE

File: mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp
Comment (cxy-1993) [reply mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp]: DONE

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (ftynse) [inline mlir/include/mlir/Interfaces/TilingInterface.td:111]: I see that this is cargo-culted from the method above, but we should be using `SmallVectorImpl` to avoid hardcoding the implicit number of stack elements. `SmallVector` is in fact `SmallVector<6>` in this case.

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (ftynse) [inline mlir/include/mlir/Interfaces/TilingInterface.td:104]: Nit: prefer globally-qualified types in .td, `::mlir::LogicalResult`. Generated code may be #included in another namespace.

File: mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp:141]: Nit: expand `auto` unless the type is obvious from line-level context, e.g., there is a cast such as below.

File: mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp:150]:
```suggestion
      for (auto &&[index, value] : llvm::enumerate(iterationDomain)) {
        mappedOffsets[index] = value.offset;
        mappedSizes[index] = value.size;
```

File: mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp:153]: Same as above.

File: mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp:177]: Nit: `emitOpError` is for errors that are related to the validity of the op, which this is not. Use `emitError` instead.

**Review by ftynse — COMMENTED** [review]
This needs tests.

File: mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp
Comment (cxy-1993) [reply mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp]: done

File: mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp
Comment (cxy-1993) [reply mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp]: done

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (cxy-1993) [reply mlir/include/mlir/Interfaces/TilingInterface.td]: Agree. I changed parameter type to SmallVectorImpl as well as the code I  cargo-culted from.

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (cxy-1993) [reply mlir/include/mlir/Interfaces/TilingInterface.td]: nice catch, done

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (cxy-1993) [reply mlir/include/mlir/Interfaces/TilingInterface.td]: done

**Comment by cxy-1993** [comment]
> This needs tests.

Thanks for your time. I have modified the code base on your comments(as well as the original code).

As for tests, I have original implemented a transform op in linalg dialect, and add tests. After the discussion above, we decided to only add the interface part for now. The tests for these parts require the implementation of the transform op, so can we implement the interface first and add tests for the transform op later?

**Comment by MaheshRavishankar** [comment]
@Abhishek-Varma  please take a look at this PR and approve if this looks fine to you and you can land your PR on top of this. 

For me this looks fine..

File: mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp
Comment (Abhishek-Varma) [inline mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp:162]: nit: Triple doc comments: `//` -> `///`.

File: mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp
Comment (Abhishek-Varma) [inline mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp:222]: Use `emitError` here.

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (Abhishek-Varma) [inline mlir/include/mlir/Interfaces/TilingInterface.td:177]: Would this require the global namespace thingy too on `TilingResult` ?

**Review by Abhishek-Varma — APPROVED** [review]
Thanks for this @cxy-1993 - this has been a really good nudge for fusing consumers!

A few comments to address before merging. LGTM otherwise.

**Comment by cxy-1993** [comment]
> Thanks for this @cxy-1993 - this has been a really good nudge for fusing consumers!
> 
> A few comments to address before merging. LGTM otherwise.

Thanks for the comment @Abhishek-Varma. The code have modified according to your comment. I don't have write access to this repo, please help me merge this patch into master.

File: mlir/include/mlir/Interfaces/TilingInterface.td
Comment (cxy-1993) [reply mlir/include/mlir/Interfaces/TilingInterface.td]: done

File: mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp
Comment (cxy-1993) [reply mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp]: done

File: mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp
Comment (cxy-1993) [reply mlir/lib/Dialect/Linalg/Transforms/TilingInterfaceImpl.cpp]: Nice catch. I have done modify it(also the original code in this file).

**Comment by nicolasvasilache** [comment]
@Abhishek-Varma, @cxy-1993, @ftynse gave a clear review that this needs tests.

It is not acceptable to land randomly untested code!

Please revert, add tests and resend for review.

**Comment by MaheshRavishankar** [comment]
Definitely needs tests, but I didn't see a need for revert. Should be fine to add tests after the fact. @cxy-1993 and @Abhishek-Varma could we add some tests here

**Comment by ftynse** [comment]
For practicality reasons, I'd prefer a revert and a new PR that includes tests as opposed to just adding tests. This will help us, the reviewers, ascertain that the tests are actually testing what they should be testing, there is enough coverage, etc. It also makes it easy to bisect and revert the change should the need arise.

**Comment by MaheshRavishankar** [comment]
We can't block code progress because of that. I have been reviewing all the code and have been keeping track of all the changes. This is also a strict addition that isn't part of what needs to be undone from the legacy code. 

It is still on my plate to address this. Everyone has time commitments that they have to balance. I will get to it in short order, but we can't hold things hostage because of that.

**Comment by stellaraccident** [comment]
> > Definitely needs tests, but I didn't see a need for revert. Should be fine to add tests after the fact. @cxy-1993 and @Abhishek-Varma could we add some tests here
> 
> yeah, let's instead be consistent with what we've been doing so far: https://mlir.llvm.org/getting_started/TestingGuide/

Let's keep the bar high. Alex's suggestion is the right one: revert and land with appropriate tests.

> 
> Additionally, I believe this should also not have landed for the same reason as I am objecting to #88712 I'll just copy paste here:
> 
> Now, I am afraid I cannot subscribe to any large code changes to tiling transforms until #77874 is addressed to my satisfaction. I already gave a pass 6 months ago and another one 3 months ago when things were supposed to be addressed "in short order".
> 
> We're now past tech-debt reduction time: first let's address [mlir][TilingInterface] Use LoopLikeOpInterface in tiling using SCF to unify tiling with scf.for and scf.forall. #77874,

Can we get a f2f to talk about this separately? I've had my eye on this for a long time too, and there is a significant hairball that (believe it or not) is being undone. It is just hard and taking more iteration than expected.

**Comment by cxy-1993** [comment]
New PR: https://github.com/llvm/llvm-project/pull/89893, just add tests compare to this version. Please help me review it, thanks. @MaheshRavishankar @ftynse @Abhishek-Varma  @nicolasvasilache

**Pull request merged**
